pkg/pullrequest/scm.go

```diff
@@ -0,0 +1,133 @@
+/*
+Copyright 2019 The Tekton Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pullrequest
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+
+	"golang.org/x/oauth2"
+
+	"github.com/jenkins-x/go-scm/scm/driver/github"
+	"github.com/jenkins-x/go-scm/scm/driver/gitlab"
+	"go.uber.org/zap"
+)
+
+func NewSCMHandler(logger *zap.SugaredLogger, raw, provider, token string) (*Handler, error) {
+	u, err := url.Parse(raw)
+	if err != nil {
+		return nil, err
+	}
+
+	if provider == "" {
+		p, err := guessProvider(raw)
+		if err != nil {
+			return nil, err
+		}
+		provider = p
+	}
+
+	var handler *Handler
+	switch provider {
+	case "github":
+		handler, err = githubHandlerFromURL(u, token, logger)
+	case "gitlab":
+		handler, err = gitlabHandlerFromURL(u, token, logger)
+	default:
+		return nil, fmt.Errorf("unsupported pr url: %s", raw)
+	}
+	return handler, err
+}
+
+func githubHandlerFromURL(u *url.URL, token string, logger *zap.SugaredLogger) (*Handler, error) {
+	split := strings.Split(u.Path, "/")
+	if len(split) < 5 {
+		return nil, fmt.Errorf("could not determine PR from URL: %v", u)
+	}
+	owner, repo, pr := split[1], split[2], split[4]
```

Comment:
Something I've started getting into the habit of doing is to annotate the logger with keywords to make it easier to filter in logs later. (https://godoc.org/go.uber.org/zap#Field)

I think it'd be useful to include: resource type ("pullrequest"), provider, owner, repo, and PR number.

Reply:
Nice!